In [160]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

In [161]:
df = pd.read_csv('marketing_campaign.csv', sep="\t")
df.head()

ID  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0  5524        1957  Graduation         Single  58138.0        0         0   
1  2174        1954  Graduation         Single  46344.0        1         1   
2  4141        1965  Graduation       Together  71613.0        0         0   
3  6182        1984  Graduation       Together  26646.0        1         0   
4  5324        1981         PhD        Married  58293.0        1         0   

  Dt_Customer  Recency  MntWines  ...  NumWebVisitsMonth  AcceptedCmp3  \
0  04-09-2012       58       635  ...                  7             0   
1  08-03-2014       38        11  ...                  5             0   
2  21-08-2013       26       426  ...                  4             0   
3  10-02-2014       26        11  ...                  6             0   
4  19-01-2014       94       173  ...                  5             0   

   AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  Complain  \
0             0             0             0             0         0   
1             0             0             0             0         0   
2             0             0             0             0         0   
3             0             0             0             0         0   
4             0             0             0             0         0   

   Z_CostContact  Z_Revenue  Response  
0              3         11         1  
1              3         11         0  
2              3         11         0  
3              3         11         0  
4              3         11         0  

[5 rows x 29 columns]

In [162]:
df.columns

Index(['ID', 'Year_Birth', 'Education', 'Marital_Status', 'Income', 'Kidhome',
       'Teenhome', 'Dt_Customer', 'Recency', 'MntWines', 'MntFruits',
       'MntMeatProducts', 'MntFishProducts', 'MntSweetProducts',
       'MntGoldProds', 'NumDealsPurchases', 'NumWebPurchases',
       'NumCatalogPurchases', 'NumStorePurchases', 'NumWebVisitsMonth',
       'AcceptedCmp3', 'AcceptedCmp4', 'AcceptedCmp5', 'AcceptedCmp1',
       'AcceptedCmp2', 'Complain', 'Z_CostContact', 'Z_Revenue', 'Response'],
      dtype='object')

In [163]:
df.shape

(2240, 29)

In [164]:
df.isnull().sum()

ID                      0
Year_Birth              0
Education               0
Marital_Status          0
Income                 24
Kidhome                 0
Teenhome                0
Dt_Customer             0
Recency                 0
MntWines                0
MntFruits               0
MntMeatProducts         0
MntFishProducts         0
MntSweetProducts        0
MntGoldProds            0
NumDealsPurchases       0
NumWebPurchases         0
NumCatalogPurchases     0
NumStorePurchases       0
NumWebVisitsMonth       0
AcceptedCmp3            0
AcceptedCmp4            0
AcceptedCmp5            0
AcceptedCmp1            0
AcceptedCmp2            0
Complain                0
Z_CostContact           0
Z_Revenue               0
Response                0
dtype: int64

In [165]:
income_mean = df.groupby('Education')['Income'].mean()

df['Income'] = df['Income'].fillna(df['Education'].map(income_mean))

In [166]:
df['Education'].map(income_mean)

0       52720.373656
1       52720.373656
2       52720.373656
3       52720.373656
4       56145.313929
            ...     
2235    52720.373656
2236    56145.313929
2237    52720.373656
2238    52917.534247
2239    56145.313929
Name: Education, Length: 2240, dtype: float64

In [167]:
df.isnull().sum()

ID                     0
Year_Birth             0
Education              0
Marital_Status         0
Income                 0
Kidhome                0
Teenhome               0
Dt_Customer            0
Recency                0
MntWines               0
MntFruits              0
MntMeatProducts        0
MntFishProducts        0
MntSweetProducts       0
MntGoldProds           0
NumDealsPurchases      0
NumWebPurchases        0
NumCatalogPurchases    0
NumStorePurchases      0
NumWebVisitsMonth      0
AcceptedCmp3           0
AcceptedCmp4           0
AcceptedCmp5           0
AcceptedCmp1           0
AcceptedCmp2           0
Complain               0
Z_CostContact          0
Z_Revenue              0
Response               0
dtype: int64

#### We will perform feature engineering : 

In [168]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2240 entries, 0 to 2239
Data columns (total 29 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   2240 non-null   int64  
 1   Year_Birth           2240 non-null   int64  
 2   Education            2240 non-null   object 
 3   Marital_Status       2240 non-null   object 
 4   Income               2240 non-null   float64
 5   Kidhome              2240 non-null   int64  
 6   Teenhome             2240 non-null   int64  
 7   Dt_Customer          2240 non-null   object 
 8   Recency              2240 non-null   int64  
 9   MntWines             2240 non-null   int64  
 10  MntFruits            2240 non-null   int64  
 11  MntMeatProducts      2240 non-null   int64  
 12  MntFishProducts      2240 non-null   int64  
 13  MntSweetProducts     2240 non-null   int64  
 14  MntGoldProds         2240 non-null   int64  
 15  NumDealsPurchases    2240 non-null   i

In [169]:
df['Education'].unique()

array(['Graduation', 'PhD', 'Master', 'Basic', '2n Cycle'], dtype=object)

In [170]:
df['Marital_Status'].unique()

array(['Single', 'Together', 'Married', 'Divorced', 'Widow', 'Alone',
       'Absurd', 'YOLO'], dtype=object)

In [171]:
df.columns

Index(['ID', 'Year_Birth', 'Education', 'Marital_Status', 'Income', 'Kidhome',
       'Teenhome', 'Dt_Customer', 'Recency', 'MntWines', 'MntFruits',
       'MntMeatProducts', 'MntFishProducts', 'MntSweetProducts',
       'MntGoldProds', 'NumDealsPurchases', 'NumWebPurchases',
       'NumCatalogPurchases', 'NumStorePurchases', 'NumWebVisitsMonth',
       'AcceptedCmp3', 'AcceptedCmp4', 'AcceptedCmp5', 'AcceptedCmp1',
       'AcceptedCmp2', 'Complain', 'Z_CostContact', 'Z_Revenue', 'Response'],
      dtype='object')

we can merge both the number of kids and teens in one column. and the number of purchases in another column 

In [172]:
from datetime import datetime

In [173]:
df.columns

Index(['ID', 'Year_Birth', 'Education', 'Marital_Status', 'Income', 'Kidhome',
       'Teenhome', 'Dt_Customer', 'Recency', 'MntWines', 'MntFruits',
       'MntMeatProducts', 'MntFishProducts', 'MntSweetProducts',
       'MntGoldProds', 'NumDealsPurchases', 'NumWebPurchases',
       'NumCatalogPurchases', 'NumStorePurchases', 'NumWebVisitsMonth',
       'AcceptedCmp3', 'AcceptedCmp4', 'AcceptedCmp5', 'AcceptedCmp1',
       'AcceptedCmp2', 'Complain', 'Z_CostContact', 'Z_Revenue', 'Response'],
      dtype='object')

In [174]:
df['children'] = df['Kidhome'] + df['Teenhome']
df.drop(columns=['Recency'], inplace=True)
df['Dt_Customer'] = pd.to_datetime(df['Dt_Customer'], format = '%d-%m-%Y')
max_date = df['Dt_Customer'].max()
df['Days_since_enrolment'] = df['Dt_Customer'].apply(lambda x: (max_date - x))
df['Age'] = datetime.now().year - df['Year_Birth']


df['Total Spending'] = df['MntWines'] + df['MntFruits'] + df['MntMeatProducts'] + df['MntFishProducts'] + df['MntSweetProducts'] + df['MntGoldProds']

df['Prob_wines'] = df['MntWines'] / df['Total Spending']
df['Prob_fruits'] = df['MntFruits'] / df['Total Spending']
df['Prob_meatproducts'] = df['MntMeatProducts'] / df['Total Spending']
df['Prob_fishproducts'] = df['MntFishProducts'] / df['Total Spending']
df['Prob_sweetproducts'] = df['MntSweetProducts'] / df['Total Spending']
df['Prop_goldprods']=df['MntGoldProds'] / df['Total Spending']


df['Offers_Responded_To'] = df['AcceptedCmp1'] + df['AcceptedCmp2'] + df['AcceptedCmp3'] + df['AcceptedCmp4'] + df['AcceptedCmp5'] + df['Response']

df['Total Purchases'] = df['NumDealsPurchases'] + df['NumWebPurchases'] + df['NumCatalogPurchases'] + df['NumStorePurchases']

df.drop(columns=['Z_CostContact', 'Z_Revenue', 'Year_Birth','Teenhome','Response', 'MntWines', 'MntFruits', 'MntMeatProducts', 'MntFishProducts',
           'MntSweetProducts', 'MntGoldProds','AcceptedCmp3', 'AcceptedCmp4', 'AcceptedCmp5', 'AcceptedCmp1', 'AcceptedCmp2', 'NumDealsPurchases', 'NumWebPurchases',
           'NumCatalogPurchases', 'NumStorePurchases','Year_Birth','Dt_Customer','ID'], inplace=True)


In [175]:
df.describe()

Income      Kidhome  NumWebVisitsMonth     Complain  \
count    2240.000000  2240.000000        2240.000000  2240.000000   
mean    52253.592375     0.444196           5.316518     0.009375   
std     25039.085601     0.538398           2.426645     0.096391   
min      1730.000000     0.000000           0.000000     0.000000   
25%     35538.750000     0.000000           3.000000     0.000000   
50%     51609.500000     0.000000           6.000000     0.000000   
75%     68289.750000     1.000000           7.000000     0.000000   
max    666666.000000     2.000000          20.000000     1.000000   

          children         Days_since_enrolment          Age  Total Spending  \
count  2240.000000                         2240  2240.000000     2240.000000   
mean      0.950446  353 days 13:58:17.142857144    54.194196      605.798214   
std       0.751803  202 days 02:56:25.066913892    11.984069      602.249288   
min       0.000000              0 days 00:00:00    27.000000        5.000000   
25%       0.000000            180 days 18:00:00    46.000000       68.750000   
50%       1.000000            355 days 12:00:00    53.000000      396.000000   
75%       1.000000            529 days 00:00:00    64.000000     1045.500000   
max       3.000000            699 days 00:00:00   130.000000     2525.000000   

        Prob_wines  Prob_fruits  Prob_meatproducts  Prob_fishproducts  \
count  2240.000000  2240.000000        2240.000000        2240.000000   
mean      0.458481     0.049505           0.249508           0.071532   
std       0.228411     0.055867           0.126633           0.077909   
min       0.000000     0.000000           0.000000           0.000000   
25%       0.289506     0.008924           0.156250           0.012571   
50%       0.457482     0.029840           0.233447           0.048193   
75%       0.639143     0.070237           0.328227           0.104703   
max       0.963303     0.445545           0.997110           0.590909   

       Prob_sweetproducts  Prop_goldprods  Offers_Responded_To  \
count         2240.000000     2240.000000          2240.000000   
mean             0.050777        0.120196             0.446875   
std              0.061184        0.108988             0.890543   
min              0.000000        0.000000             0.000000   
25%              0.008621        0.038095             0.000000   
50%              0.033333        0.086294             0.000000   
75%              0.070411        0.169768             1.000000   
max              0.945848        0.894150             5.000000   

       Total Purchases  
count      2240.000000  
mean         14.862054  
std           7.677173  
min           0.000000  
25%           8.000000  
50%          15.000000  
75%          21.000000  
max          44.000000

we will remove outliers such as age that has 130 years as a maximum and we will remove outliers from income.

In [176]:

columns = ['Income', 'Age','Total Purchases']
for column in columns:
  Q1 = df[column].quantile(0.1)
  Q3 = df[column].quantile(0.9)
  IQR = Q3 - Q1 
  lower_bound = Q1 - 1.5 * IQR
  upper_bound = Q3 + 1.5 * IQR
  df = df[(df[column] >= lower_bound) & (df[column] <= upper_bound) ]


In [177]:
df.shape

(2236, 18)

In [178]:
df.describe()

Income      Kidhome  NumWebVisitsMonth     Complain  \
count    2236.000000  2236.000000        2236.000000  2236.000000   
mean    51968.258909     0.444097           5.318873     0.008945   
std     21412.998830     0.538459           2.426886     0.094173   
min      1730.000000     0.000000           0.000000     0.000000   
25%     35502.500000     0.000000           3.000000     0.000000   
50%     51566.000000     0.000000           6.000000     0.000000   
75%     68275.750000     1.000000           7.000000     0.000000   
max    162397.000000     2.000000          20.000000     1.000000   

          children         Days_since_enrolment          Age  Total Spending  \
count  2236.000000                         2236  2236.000000     2236.000000   
mean      0.950805  353 days 18:33:29.302325580    54.101968      605.986583   
std       0.752204  202 days 04:21:26.839418520    11.703281      601.865156   
min       0.000000              0 days 00:00:00    27.000000        5.000000   
25%       0.000000            180 days 18:00:00    46.000000       69.000000   
50%       1.000000            356 days 00:00:00    53.000000      396.500000   
75%       1.000000            529 days 00:00:00    64.000000     1045.500000   
max       3.000000            699 days 00:00:00    83.000000     2525.000000   

        Prob_wines  Prob_fruits  Prob_meatproducts  Prob_fishproducts  \
count  2236.000000  2236.000000        2236.000000        2236.000000   
mean      0.458789     0.049416           0.249532           0.071387   
std       0.228457     0.055772           0.126709           0.077790   
min       0.000000     0.000000           0.000000           0.000000   
25%       0.290067     0.008924           0.156250           0.012560   
50%       0.457822     0.029774           0.233447           0.048158   
75%       0.639765     0.070080           0.328441           0.103809   
max       0.963303     0.445545           0.997110           0.590909   

       Prob_sweetproducts  Prop_goldprods  Offers_Responded_To  \
count         2236.000000     2236.000000          2236.000000   
mean             0.050818        0.120058             0.447227   
std              0.061224        0.108929             0.891113   
min              0.000000        0.000000             0.000000   
25%              0.008621        0.038081             0.000000   
50%              0.033333        0.085714             0.000000   
75%              0.070719        0.169477             1.000000   
max              0.945848        0.894150             5.000000   

       Total Purchases  
count      2236.000000  
mean         14.872540  
std           7.677874  
min           0.000000  
25%           8.000000  
50%          15.000000  
75%          21.000000  
max          44.000000

In [179]:
df.head()

Education Marital_Status   Income  Kidhome  NumWebVisitsMonth  Complain  \
0  Graduation         Single  58138.0        0                  7         0   
1  Graduation         Single  46344.0        1                  5         0   
2  Graduation       Together  71613.0        0                  4         0   
3  Graduation       Together  26646.0        1                  6         0   
4         PhD        Married  58293.0        1                  5         0   

   children Days_since_enrolment  Age  Total Spending  Prob_wines  \
0         0             663 days   66            1617    0.392703   
1         2             113 days   69              27    0.407407   
2         0             312 days   58             776    0.548969   
3         1             139 days   39              53    0.207547   
4         1             161 days   42             422    0.409953   

   Prob_fruits  Prob_meatproducts  Prob_fishproducts  Prob_sweetproducts  \
0     0.054422           0.337662           0.106370            0.054422   
1     0.037037           0.222222           0.074074            0.037037   
2     0.063144           0.163660           0.143041            0.027062   
3     0.075472           0.377358           0.188679            0.056604   
4     0.101896           0.279621           0.109005            0.063981   

   Prop_goldprods  Offers_Responded_To  Total Purchases  
0        0.054422                    1               25  
1        0.222222                    0                6  
2        0.054124                    0               21  
3        0.094340                    0                8  
4        0.035545                    0               19

In [180]:
df.shape

(2236, 18)

Making age column into categories : 

In [181]:
bins = list(range(0, 100, 10))

labels = ['0-10','11-20','21-30','31,40','41-50', '51-60', '61-70', '71-80', '81-90']

df['Age_group'] = pd.cut(df['Age'], bins, labels=labels)

In [182]:
df.head()

Education Marital_Status   Income  Kidhome  NumWebVisitsMonth  Complain  \
0  Graduation         Single  58138.0        0                  7         0   
1  Graduation         Single  46344.0        1                  5         0   
2  Graduation       Together  71613.0        0                  4         0   
3  Graduation       Together  26646.0        1                  6         0   
4         PhD        Married  58293.0        1                  5         0   

   children Days_since_enrolment  Age  Total Spending  Prob_wines  \
0         0             663 days   66            1617    0.392703   
1         2             113 days   69              27    0.407407   
2         0             312 days   58             776    0.548969   
3         1             139 days   39              53    0.207547   
4         1             161 days   42             422    0.409953   

   Prob_fruits  Prob_meatproducts  Prob_fishproducts  Prob_sweetproducts  \
0     0.054422           0.337662           0.106370            0.054422   
1     0.037037           0.222222           0.074074            0.037037   
2     0.063144           0.163660           0.143041            0.027062   
3     0.075472           0.377358           0.188679            0.056604   
4     0.101896           0.279621           0.109005            0.063981   

   Prop_goldprods  Offers_Responded_To  Total Purchases Age_group  
0        0.054422                    1               25     61-70  
1        0.222222                    0                6     61-70  
2        0.054124                    0               21     51-60  
3        0.094340                    0                8     31,40  
4        0.035545                    0               19     41-50

In [183]:
df.describe()

Income      Kidhome  NumWebVisitsMonth     Complain  \
count    2236.000000  2236.000000        2236.000000  2236.000000   
mean    51968.258909     0.444097           5.318873     0.008945   
std     21412.998830     0.538459           2.426886     0.094173   
min      1730.000000     0.000000           0.000000     0.000000   
25%     35502.500000     0.000000           3.000000     0.000000   
50%     51566.000000     0.000000           6.000000     0.000000   
75%     68275.750000     1.000000           7.000000     0.000000   
max    162397.000000     2.000000          20.000000     1.000000   

          children         Days_since_enrolment          Age  Total Spending  \
count  2236.000000                         2236  2236.000000     2236.000000   
mean      0.950805  353 days 18:33:29.302325580    54.101968      605.986583   
std       0.752204  202 days 04:21:26.839418520    11.703281      601.865156   
min       0.000000              0 days 00:00:00    27.000000        5.000000   
25%       0.000000            180 days 18:00:00    46.000000       69.000000   
50%       1.000000            356 days 00:00:00    53.000000      396.500000   
75%       1.000000            529 days 00:00:00    64.000000     1045.500000   
max       3.000000            699 days 00:00:00    83.000000     2525.000000   

        Prob_wines  Prob_fruits  Prob_meatproducts  Prob_fishproducts  \
count  2236.000000  2236.000000        2236.000000        2236.000000   
mean      0.458789     0.049416           0.249532           0.071387   
std       0.228457     0.055772           0.126709           0.077790   
min       0.000000     0.000000           0.000000           0.000000   
25%       0.290067     0.008924           0.156250           0.012560   
50%       0.457822     0.029774           0.233447           0.048158   
75%       0.639765     0.070080           0.328441           0.103809   
max       0.963303     0.445545           0.997110           0.590909   

       Prob_sweetproducts  Prop_goldprods  Offers_Responded_To  \
count         2236.000000     2236.000000          2236.000000   
mean             0.050818        0.120058             0.447227   
std              0.061224        0.108929             0.891113   
min              0.000000        0.000000             0.000000   
25%              0.008621        0.038081             0.000000   
50%              0.033333        0.085714             0.000000   
75%              0.070719        0.169477             1.000000   
max              0.945848        0.894150             5.000000   

       Total Purchases  
count      2236.000000  
mean         14.872540  
std           7.677874  
min           0.000000  
25%           8.000000  
50%          15.000000  
75%          21.000000  
max          44.000000

In [184]:
import seaborn as sns
import plotly.express as px
import plotly.figure_factory as ff
import plotly.graph_objects as go

In [185]:
age_group = df['Age_group'].value_counts().reset_index()

age_group


Age_group  count
0     51-60    620
1     41-50    619
2     61-70    465
3     31,40    286
4     71-80    229
5     21-30     15
6     81-90      2
7      0-10      0
8     11-20      0

In [186]:
px.bar(age_group, x = 'Age_group', y='count')

The analysis reveals a substantial portion of customers falling within the age range of over 30 years. Moreover, the largest concentration is observed among customers aged over 40 but younger than 70. As a result, it would be prudent to prioritize the analysis and focus on these five age groups.

In [193]:
customers_age_ms_group = df.groupby(["Age_group", "Marital_Status"]).size()\
                                                                .reset_index(name="Count")

target_age_groups = ["31-40", "41-50", "51-60", "61-70", "71-80"]

fig = px.bar(customers_age_ms_group.loc[customers_age_ms_group.Age_group.isin(target_age_groups), :], 
             x="Age_group", y="Count", color="Marital_Status", barmode="group")
fig.show()